test/e2e/cluster_logging_utils.go

```diff
@@ -19,51 +19,198 @@ package e2e
+	return lineNumber, err == nil
+}
+
+func createLoggingPod(f *framework.Framework, podName string, totalLines int, loggingDuration time.Duration) *loggingPod {
+	framework.Logf("Starting pod %s", podName)
+	createLogsGeneratorPod(f, podName, totalLines, loggingDuration)
+
+	return &loggingPod{
+		Name: podName,
+		// It's used to avoid querying logs from before the pod was started
+		LastTimestamp:       time.Now(),
+		Occurrences:         make(map[int]*logEntry),
+		ExpectedLinesNumber: totalLines,
+	}
+}
+
+func createLogsGeneratorPod(f *framework.Framework, podName string, linesCount int, duration time.Duration) {
+	f.PodClient().Create(&api_v1.Pod{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name: podName,
 		},
-		Spec: v1.PodSpec{
-			RestartPolicy: v1.RestartPolicyOnFailure,
-			Containers: []v1.Container{
+		Spec: api_v1.PodSpec{
+			RestartPolicy: api_v1.RestartPolicyNever,
+			Containers: []api_v1.Container{
 				{
-					Name:  synthLoggerPodName,
-					Image: "gcr.io/google_containers/busybox:1.24",
-					// notice: the subshell syntax is escaped with `$$`
-					Command: []string{"/bin/sh", "-c", fmt.Sprintf("i=0; while [ $i -lt %d ]; do echo $i; i=`expr $i + 1`; done", linesCount)},
+					Name:  podName,
+					Image: "gcr.io/google_containers/logs-generator:v0.1.0",
+					Env: []api_v1.EnvVar{
+						{
+							Name:  "LOGS_GENERATOR_LINES_TOTAL",
+							Value: strconv.Itoa(linesCount),
+						},
+						{
+							Name:  "LOGS_GENERATOR_DURATION",
+							Value: duration.String(),
+						},
+					},
+					Resources: api_v1.ResourceRequirements{
+						Requests: api_v1.ResourceList{
+							api_v1.ResourceCPU: *resource.NewMilliQuantity(
+								loggingContainerCpu,
+								resource.DecimalSI),
+							api_v1.ResourceMemory: *resource.NewQuantity(
+								loggingContainerMemory,
+								resource.BinarySI),
+						},
+					},
 				},
 			},
 		},
 	})
 }
 
-func reportLogsFromFluentdPod(f *framework.Framework) error {
-	synthLoggerPod, err := f.PodClient().Get(synthLoggerPodName, metav1.GetOptions{})
+func waitForLogsIngestion(logsProvider logsProvider, pods []*loggingPod, ingestionTimeout time.Duration) error {
+	totalMissing := 0
+	for _, pod := range pods {
+		totalMissing = pod.ExpectedLinesNumber
+	}
+
+	missingByPod := make([]int, len(pods))
+	for podIdx, pod := range pods {
+		missingByPod[podIdx] = pod.ExpectedLinesNumber
+	}
+
+	for start := time.Now(); totalMissing > 0 && time.Since(start) < ingestionTimeout; time.Sleep(ingestionRetryDelay) {
+		missing := 0
+		for podIdx, pod := range pods {
+			if missingByPod[podIdx] == 0 {
+				continue
+			}
+
+			missingByPod[podIdx] = pullMissingLogsCount(logsProvider, pod)
+			missing += missingByPod[podIdx]
+		}
+
+		totalMissing = missing
+		if totalMissing > 0 {
+			framework.Logf("Still missing %d lines in total", totalMissing)
+		}
+	}
+
+	if totalMissing > 0 {
+		return fmt.Errorf("After %v still missing %d lines", ingestionTimeout, totalMissing)
```

Comment:
There are log messages which indicate how many lines are missing for the particular pod